### DISCUSSION TITLE
How to update the first record i find in the where clause? (updateFirst method maybe?)

### DISCUSSION THREAD

**Discussion by ADrejta** [prisma/prisma#5422] [Q&A] [answered]
There are cases  when you want to update only the first record you find like the findFirst method finds only the first record that matches the search pattern given.

When wanting to update a row and ensuring that it is update in a transactional matter, going through the route of first fetching the row then updating it is not really possible without locking the table.

For example:

Let's say we have a table of documents like so:
+----+------------------------+---------------+
| id   | document_path    | processed |
+----+------------------------+---------------+
| 1  | some random path |       0          |
+----+------------------------+---------------+
| 2  | some random path |       0          |
+----+------------------------+---------------+
| 3  | some random path |       1          |
+----+------------------------+---------------+

And now i have two servers that are pooling the database to get any unprocessed documents.

Okay, I'l go ahead and first fetch the first document that is unprocessed then update it like so:
```ts
const document = await prisma.documents.findFirst({
         where: {
            processed: 0,
         },
 });
 
await prisma.documents.update({
   where: { id: document.id },
   data: { processed : 1 }
 })
 ```
Sounds good until you figure out that booth servers got the same document to process XD.

The reason why that happened was that while the first server was updating the table server 2 already got the same document to process that server 1 did.

How can i read and update one row at the same time using prisma? Is there a method like findFirst but for updates? I couldn't find something like that.

What way can i achieve what I'm trying to do in this situation using prisma? 

I will apologize in advance if i have missed anything obvious in the docs , but I can't seem to find anything to fit my use case.

**Comment by ryands17** [comment, marked as answer]
@ADrejta 👋 

In this case, you would need to use Optimistic Concurrency Control to first fetch the data and only update if the `processed` matches the version that you fetched. This can be implemented as explained in the docs [here](https://www.prisma.io/docs/guides/prisma-guides/prisma-client-transactions-guide#optimistic-concurrency-control). Let me know if you have any queries :)

**Reply by ADrejta** [reply]
@ryands17 :wave: 

Wow, i feel stupid now :laughing: 

That was exactly what i was looking for and I completely skipped it while reading the docs and guides. 

So this is the way i did it:

```javascript
const document = await prisma.documents.findFirst({
   where: {
      processed: 0,
   },
});
const updatedCount = await prisma.documents.updateMany({
   where: {
      id: document.id,
      processed: 0,
   },
   data: {
      processed: -1,
   },
});
if (updatedCount.count === 0) {
   // go fetch another document since this one is already taken
}
```
I made the processed column have a "-1" value for when it is fetched and then proceed to a "1" value for when the processing is done.
So if the document was fetched beforehand by another worker server I can tell the current worker server to just go fetch another one.

Keep up the good work with prisma. I am really enjoying it so far :smiley:

**Reply by shtse8** [reply]
you need to do it in transaction to prevent racing.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
